### DISCUSSION TITLE
How to Set CUDA_VISIBLE_DEVICES Per Worker in SpecCluster for Multi-GPU Training?

### DISCUSSION THREAD

**Discussion by kwanUm** [dask/dask#11626] [Q&A]
Hi Dask community, 🙏

I'm trying to simulate the behavior of torchrun using Dask for training with dask + FSDP on 8 GPUs. In this setup, each worker needs to have its own CUDA_VISIBLE_DEVICES environment variable set before the worker process starts to ensure proper GPU isolation.

Unfortunately, I’ve noticed that CUDA_VISIBLE_DEVICES is inherited from the main script, meaning all workers share the same environment when they load. The CUDA_VISIBLE_DEVICES is set only after the main script is loading.

Is there a way to modify the SpecCluster implementation so that the os.environ can be set with the correct CUDA_VISIBLE_DEVICES for each worker before it starts? I want to ensure that each worker process has its environment configured correctly for multi-GPU training before anything is imported in it.

Any pointers or suggestions on how to achieve this, or if this is feasible with the current design, would be greatly appreciated!

**Comment by maneesh29s** [comment]
Have you tried Worker Plugins?
https://distributed.dask.org/en/latest/plugins.html#worker-plugins

**Comment by kwanUm** [comment]
Yes, however you may register a single worker plugin per all workers, was
non-trivial for me to hack it around to alloc different device for each
worker.

**Reply by maneesh29s** [reply]
Yeah, the plugin code is same across all workers.

I would have created a [distributed variable](https://distributed.dask.org/en/stable/api.html#distributed.Variable) in my client code, which contains an array of  CUDA_VISIBLE_DEVICES values, 
then in my worker plugin I would select on of the values based on some index (not sure exactly how it will work).
